### PR TITLE
[Owners Review] Added CAN example for XNET

### PR DESCRIPTION
**Pull request by kothariaman**
### What does this Pull Request accomplish?
Implemented CAN examples for XNET. Ported and replicated them from C to python

### Why should this Pull Request be merged?
Existing C examples are ported to Python using gRPC API's through the gRPC server

### What testing has been done?
Locally tested the python version and matched the output with the C API.
It reports a grpc error when server is not running
image

### CODE REVIEW

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:1]: Add a comment stating what this example actually does and some additional information on getting started. you can see what other have done.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:77]: This can be simplified to `i = i+1 if i <= 10 else 0`

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:25]: Not sure, if this blank line is required. Probably we can remove it and it would still look readable. Ill leave that to you.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:21]: Spaces after `,` while passing arguments to `StatusToString`. Something like `nixnet_types.StatustoString(Status, status_string.__sizeof__(), Source)`

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:19]: Shouldn't the function name be `display_error_and_exit` ? Is this name used in the C Example? Anyways we can still change it to make it more clear and readable.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:18]: Since this a local function, the description of this function should come inside the body. Something like this.

```
def display_error_and_exit(Status, Source)::
"""Display error message and exit."""
... rest of the code
```

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:73]: Here we can use `f` strings. Something like this `print(f"Signal 1: {value_buffer[0]}")`

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:74]: Similarly here. `print(f"Signal 2: {value_Buffer[1]\n}")`

**Comment by ni-pkhare** [comment]
The python build failures can be corrected by installing ni-python-style-guide and black. I dont remember how to do that. You could check it on wiki of grpc-device channel in Teams.

**Review by ni-pkhare — APPROVED** [review]

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:9]: We should also add a logic to take Server_adress and server_port as command line arguments from the clients. See other examples.

Comment (kothariaman) [reply examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:41]: Added this logic

Comment (doshirohan) [reply examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:41]: `pip install ni-python-styleguide` and `pip install black`. 

To execute them:

> ni-python-styleguide --exclude .\source\codegen\metadata\ lint .\source\codegen\ .\examples\

This will show you any errors flagged by ni styleguide

> black source/codegen examples

This command will report any issues flagged by black and fix it as well.

Here's the wiki that has some relevant information
[grpc-device (Teams) -> wiki](https://teams.microsoft.com/l/entity/com.microsoft.teamspace.tab.wiki/tab::7426a53d-3098-4c17-9bb0-5ba77dc51c7f?context=%7B%22subEntityId%22%3A%22%7B%5C%22pageId%5C%22%3A7%2C%5C%22origin%5C%22%3A2%7D%22%2C%22channelId%22%3A%2219%3A658a48f1bcd240dc9c56f5eb35761898%40thread.tacv2%22%7D&tenantId=87ba1f9a-44cd-43a6-b008-6fdb45a5204e)

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:10]: Everywhere it says NI-DCPOWER, pl change it accordingly.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (ni-pkhare) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:24]: Change it to use XNET example, right now this is the dcpower example

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:1]: Change file name to follow conventions for python.

Maybe `can-signal-single-point-output.py`?

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:8]: Documentation is installed at
> C:\Users\Public\Documents\National Instruments\NI-XNET\Documentation\NI-XNET Manual.chm

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:25]: Update command once you rename the file.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:23]: There is no "resource name" parameter in this case. Fix this line.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:59]: Doesn't look like this line is going to wait here for user input. So this message doesn't make sense. Make these changes to keep consistency with other gRPC examples we ship out of this repo.

- I think instead of printing error and exiting, you can `raise Exception(response.status_description)` here. That will also take care of printing call stack, so users would know which line resulted in error.

- Also, put main body of this file in try-except-finally block. `except` can catch RPC errors and print RPC specific error message. `finally` block can be used for closing session if  `session_reference.id` is not zero (id != 0 makes sure that error wasn't during session initialization and we should close the sesssion).

- Rename this method to check_for_error and check for `status != SUCCESS` here to raise exception conditionally. Callers below can call this API unconditionally after each gRPC calls and print messages you have in success scenarios after the `check_for_error`. Those won't be executed in case of errors because of `raise Exception` here. 

Look at other examples to refer to this pattern.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:68]: ~NI-DCPower~ => NI-XNET

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:65]: You don't really need to declare `session_reference` variable in advance.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:77]: Change variable names to follow lower_snake_case convention.

`create_session_response`

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:101]: Every gRPC call is going to have a response. Let's name this based on which API this is for. `write_signal_response`. Same for the `Clear` call below as well.

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:108]: Do you want to add newline after each of these print statements or does it look good as is?

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [inline examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:115]: Clears => Clear
X-NET => XNET (it doesn't refer to network of x-men ;P )

**Review by doshirohan — APPROVED** [review]

File: examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py
Comment (doshirohan) [reply examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:108]: Put screenshot of output in your PR description. It would easier for Harsha as well to visualize the output.

Comment (kothariaman) [reply examples/nixnet/CAN_SIGNAL_SINGLE_POINT_OUTPUT.py:108]: It look good as it is currently. I will update the screenshot in PR description.

**Review by Akshata440 — APPROVED** [review]

**Comment by astarche** [comment]
Both `black` and `ni-python-styleguide` are in `python_build_requirements.txt`. One way to get those is to active the venv that is created in `build \venv`. When you are editing in grpc-device with vscode, you should select that as your venv, which will give you those tools in your command prompt etc. The wiki does have instructions for how to configure black in vscode.

`validate_examples.py` will install a lot of stuff in the poetry project that it creates for validating examples. It actually doesn't include `ni-python-styleguide`, but it could.

I don't think we've defined a best practice for how to run ni-python-styleguide during development. It's not quite as straightforward as configuring black in your IDE. You should be able to enable it via `flake8` ide integration using our special ruleset, but I don't have a step-by-step for how to do that. But *you can* run it from the command line when you are in a venv with in installed, which is generally what I have been doing. FYI @DavidCurtiss

File: examples/nixnet/can-signal-single-point-output.py
Comment (astarche) [inline examples/nixnet/can-signal-single-point-output.py:46]: Consider; We usually allow the device name to be passed in as a command line argument as well. Does that work in this case? Or are there just too many names that need to match your MAX config?

File: examples/nixnet/can-signal-single-point-output.py
Comment (astarche) [inline examples/nixnet/can-signal-single-point-output.py:104]: Check for understanding: you say in the PR that the output is the same as the C API. But this is only printing data from the write buffer. So all that validates (for now) is that write didn't return an error?

File: examples/nixnet/can-signal-single-point-output.py
Comment (astarche) [inline examples/nixnet/can-signal-single-point-output.py:120]: I'm pretty sure that this can all just be `if session_reference:`.

Also make sure that you include some error cases in your testing (and include that in the PR notes).

File: examples/nixnet/can-signal-single-point-output.py
Comment (astarche) [inline examples/nixnet/can-signal-single-point-output.py:99]: nit: I don't like this long name `session_reference` and it's going to be all over the place. What about just `session`?

Also, @doshirohan have y'all considered changing `session_ref` to `session`? I think that came up in a previous PR.

**Review by astarche — APPROVED** [review]

File: examples/nixnet/can-signal-single-point-output.py
Comment (astarche) [inline examples/nixnet/can-signal-single-point-output.py:31]: Consider: Do we really need an external dependency for `q`-to-quit? There are several examples that use `ctrl+c` and maybe one that uses esc. Will one of those work for you instead? Can you require <enter> for a newline instead of a character? Can you just run 10 times without user input?

Also, if you add this, you need to add it to validate_examples as a dependency to install.

**Comment by astarche** [comment]
The current failure is from `validate_examples` which you can run locally. It is complaining about methods that are not defined (are you using something that isn't in the main proto yet?). And also about a missing dependency (which I mentioned in a comment; you should remove. but, if you do need an external dep, you can add it to validate_examples in the `poetry add` statement).

File: examples/nixnet/can-signal-single-point-output.py
Comment (kothariaman) [reply examples/nixnet/can-signal-single-point-output.py:46]: Yeah, we have implemented the same as passing the device name as interface in command line argument.

File: examples/nixnet/can-signal-single-point-output.py
Comment (kothariaman) [reply examples/nixnet/can-signal-single-point-output.py:104]: No, the function `check_for_error` raises a exception whenever there is a error and the print statements will be executed only if the status doesn't indicate a error.

File: examples/nixnet/can-signal-single-point-output.py
Comment (kothariaman) [reply examples/nixnet/can-signal-single-point-output.py:120]: Sure

File: examples/nixnet/can-signal-single-point-output.py
Comment (kothariaman) [reply examples/nixnet/can-signal-single-point-output.py:31]: Yeah, so we are going ahead with running this 10 times without user input and printing output. Removing the `getch` dependency.

File: examples/nixnet/can-signal-single-point-output.py
Comment (kothariaman) [reply examples/nixnet/can-signal-single-point-output.py:99]: Renaming this to session.

File: examples/nixnet/can-signal-single-point-output.py
Comment (astarche) [reply examples/nixnet/can-signal-single-point-output.py:104]: That's what I was thinking, sorry if my message was not clear.

That sounds fine.

Comment (kothariaman) [reply examples/nixnet/can-signal-single-point-output.py:104]: @astarche  I have reset your vote. Could you please review and approve the changes once again?

**Review by astarche — APPROVED** [review]

File: examples/nixnet/can-signal-single-point-output.py
Comment (doshirohan) [reply examples/nixnet/can-signal-single-point-output.py:99]: I just noticed this comment. @astarche we do have a task to address your earlier PR comment to fix session reference param names (and also enum names that you had called out on other PR). We will be fixing up examples as well during that change.

File: examples/nixnet/can-signal-single-point-output.py
Comment (doshirohan) [inline examples/nixnet/can-signal-single-point-output.py:1]: There seems to be an empty `PortExamplesC-Python` file above. I couldn't add a comment there, so adding it here. Make sure remove that file before you complete the PR.

File: examples/nixnet/can-signal-single-point-output.py
Comment (doshirohan) [inline examples/nixnet/can-signal-single-point-output.py:3]: >  writes a signal value when a keyboard character

This behavior has changed. Make sure to update the comment.

**Review by reckenro — APPROVED** [review]

**Pull request merged**
